lib/appsignal/utils/map_filter.ex

```diff
@@ -2,14 +2,58 @@ defmodule Appsignal.Utils.MapFilter do
   @moduledoc false
   require Logger
 
-  # Phoenix parameter filtering copied from to ensure we support the same filtering options:
-  # https://github.com/phoenixframework/phoenix/blob/dfb0c00d2077e10f8df6cc6e334e04924c4c2bcd/lib/phoenix/logger.ex#L157-L199
+  # Phoenix parameter filtering adapted from the Phoenix Framework.
+  # Copyright (c) 2014 Chris McCord - Licensed under MIT
+  #
+  # Phoenix < 1.18: https://github.com/phoenixframework/phoenix/blob/dfb0c00d2077e10f8df6cc6e334e04924c4c2bcd/lib/phoenix/logger.ex#L157-L199
+  # Phoenix >= 1.18: https://github.com/phoenixframework/phoenix/blob/8a6baa5e2ddc9cf7a2fc797ac907c40389139122/lib/phoenix/logger.ex#L180-L228
 
-  def filter(values, params \\ Application.get_env(:phoenix, :filter_parameters, []))
-  def filter(values, {:discard, params}), do: discard_values(values, params)
-  def filter(values, {:keep, params}), do: keep_values(values, params)
-  def filter(values, params), do: discard_values(values, params)
+  def filter(values, filter \\ Application.get_env(:phoenix, :filter_parameters, [])) do
+    case filter do
+      # Phoenix >= 1.18
+      {:compiled, key_match, value_match} ->
+        discard_values(values, key_match, value_match)
 
+      # Phoenix < 1.18
+      {:discard, params} ->
+        discard_values(values, params)
+
+      {:keep, match} ->
+        keep_values(values, match)
+
+      # Phoenix < 1.18
+      params ->
+        discard_values(values, params)
+    end
+  end
+
+  # Phoenix >= 1.18
+  defp discard_values(%{__struct__: mod} = struct, _key_match, _value_match) when is_atom(mod) do
+    struct
+  end
+
+  defp discard_values(%{} = map, key_match, value_match) do
+    Enum.into(map, %{}, fn {k, v} ->
+      cond do
+        is_binary(k) and String.contains?(k, key_match) ->
+          {k, "[FILTERED]"}
+
+        is_binary(v) and String.contains?(v, value_match) ->
+          {k, "[FILTERED]"}
```

Comment:
Took me a second to square how we use it (with `String.contains?/2`) with [the reported shape of the error in the original issue](https://github.com/appsignal/appsignal-elixir/issues/1008#issuecomment-3082761818). Turns out this is [the `:binary.cp()` data type from Erlang](https://www.erlang.org/doc/apps/stdlib/binary.html#t:cp/0), representing a pattern to match against. TIL!